### ISSUE TITLE
Creation of very big PNG Images is slow

### ISSUE THREAD

**Issue by ax3l** [ComputationalRadiationPhysics/picongpu#693]
As reported by @psychocoderHPC , the creation of very big PNGs (~10 Mega Pixels and above) is very slow.

- [x] **verified:**
A simple work-around for that, since PNGs are only a preview anyway, is to set [scale_image](https://github.com/ComputationalRadiationPhysics/picongpu/blob/a7e0fcda6c7b68b6e275086574c0df7ea5fd38c4/src/picongpu/include/simulation_defines/param/visualization.param#L31) in `visualization.param` to a smaller number to scale down the image that needs to be written.

- [x] **verified:** #698
We should [add](https://github.com/ComputationalRadiationPhysics/picongpu/blob/a7e0fcda6c7b68b6e275086574c0df7ea5fd38c4/src/picongpu/include/plugins/output/images/PngCreator.hpp#L101-L102) `png.setcompressionlevel(1);` to get a 2.3x write speedup.

- [ ] **open question:**
We should check if we actually [plot()](https://github.com/ComputationalRadiationPhysics/picongpu/blob/a7e0fcda6c7b68b6e275086574c0df7ea5fd38c4/src/picongpu/include/plugins/output/images/PngCreator.hpp#L104-L111) according to the internal data structure of pngwriter - maybe switching the indexes also has an influence (could require to switch the values in data, too).

- [x] **tuned**:
https://github.com/pngwriter/pngwriter/pull/46 is faster with `memset` in "black" image constructors

- [ ] **open question:**
One could check if the [zlib options](http://www.zlib.net/manual.html#Advanced) for internal memory can be set to a value that is not totally 90s - via libpng / https://github.com/pngwriter/pngwriter/issues/47 ;)

Right now we are investigating if the problem is caused by [libpng](http://www.libpng.org/pub/png/libpng.html) itself or by [pngwriter](https://github.com/pngwriter/pngwriter).

**Comment by ax3l** [comment]
Measurements locally on my SSD / laptop with diamond example
```c++
#include <pngwriter.h>
#include <cstdlib>

int main( )
{
  int w = 10000;
  int h = 10000;

  pngwriter img( w, h, 0, "/tmp/foo.png" );

  img.diamond      ( w*0.25, h/2, h/2, h/2,  1.0, 1.0, 1.0 );
  img.filleddiamond( w*0.75, h/2, h/2, h/2,  1.0, 1.0, 1.0 );
  img.write_png();

  //system( "convert /tmp/foo.png /tmp/foo.xpm; cat /tmp/foo.xpm" );
  return 0;
}
```

|`w`*`h`   | time    | size [Byte] |
|-----------|----------|--------|
| 1e2^2   | 6ms     | 4K    |
| 1e3^2   | 122ms | 16K  |
| 1e4^2   | 10s      | 668K |
||-----------|----------|-------|

**Comment by ax3l** [comment]
Random write test:

```bash
CXXFLAGS="-std=c++11" cmake ~/proggen_partiell/pngwriter/ && make
time ./diamond && du -hs /tmp/foo.png && display /tmp/foo.png
```

```c++
#include <pngwriter.h>

#include <stdlib.h>     /* srand, rand */
#include <time.h>       /* time */

#include <iostream>
#include <chrono>

using namespace std;
using namespace std::chrono;

int main( )
{
  int w = 1000;
  int h = 1000;

  srand(time(NULL));
  typedef high_resolution_clock::time_point TV;

  pngwriter img( w, h, 0, "/tmp/foo.png" );

  TV t1 = high_resolution_clock::now();
  double val = 0.0;
  const double maxVal = 65535.;
  for( int x = 0; x < w; ++x )
  {
    for( int y = 0; y < h; ++y )
    {
       val = maxVal * 0.1 / ( rand() % 10 );
       img.plot(x, y,int(val),int(val),int(val));
    }
  }
  TV t2 = high_resolution_clock::now();

  cout << "random value time: "
       << std::chrono::duration_cast<std::chrono::microseconds>( t2 - t1 ).count()
       << std::endl;

  t1 = high_resolution_clock::now();
  img.write_png();
  t2 = high_resolution_clock::now();

  cout << "write_png() time: "
       << std::chrono::duration_cast<std::chrono::microseconds>( t2 - t1 ).count()
       << std::endl;

  return 0;
}
```

| `w`*`h`   | total [s]  | init rand [s] | write png [s] |size [Byte]    |
|------------|-----------|----------------|-----------------|----------------|
|1e2^2     | 0.007    | <0.001        | 0.005            | 20 K           |
|2e2^2     | 0.02      | 0.002          | 0.017            | 76 K           |
|4e2^2     | 0.08      | 0.009          | 0.071            | 292 K         |
|7e2^2     | 0.26      | 0.026          | 0.213            | 884 K         |
|1e3^2     | 0.5        | 0.043          | 0.45              | 1.8 M          |
|4e3^2     | 8           | 0.777          | 6.99              | 30 M          |
|8e3^2     | 32         | 2.8              | 27.9              | 120 M         |

**Comment by ax3l** [comment]
Added a fast-init with `memset` in https://github.com/pngwriter/pngwriter/issues/46 to `pngwriter`.

In our project, both `tools/png2gas` and `plugins/output/images/PngCreator.hpp` use `0` (black) as the default init value, so the speedup applies for them.

**Comment by ax3l** [comment]
The default compression level seems to be `disabled`, too.
I [benchmarked](https://github.com/pngwriter/pngwriter/issues/47#issuecomment-73949513) a bit with random data and using e.g. the lowest compression level `0` increases throughput by a factor 2:

```c++
img.setcompressionlevel(0);
```

**Comment by ax3l** [comment]
- [ ] we should check if we actually [plot()](https://github.com/ComputationalRadiationPhysics/picongpu/blob/a7e0fcda6c7b68b6e275086574c0df7ea5fd38c4/src/picongpu/include/plugins/output/images/PngCreator.hpp#L104-L111) according to the internal data structure of pngwriter - maybe switching the indexes also has an influence (could require to switch the values in `data`, too).

**Comment by ax3l** [comment]
Closed due to export to threads in #995 and due to the major time and memory concern that arrives from #1279
